### PR TITLE
Improving vectorized mode 

### PR DESCRIPTION
**Pull request by xukai92**
This PR improves vectorized mode
- [x] Compatibility of `MultinomialTS`
- [x] Revert of phase points
- [x] Bugfixes for some utilites like `show`

### CODE REVIEW

File: src/trajectory.jl [hunk 1]
```diff
@@ -227,22 +227,52 @@ samplecand(rng, τ::StaticTrajectory{EndPointTS}, h, z) = step(τ.integrator, h,
...
     i = 0
     while cp < u
-        cp += p[i +=1]
+        cp += p[i+=1]
     end
-    return xs[max(i, 1)]
+    return max(i, 1)
+end
+
+function randcat(rng, P_unorm::AbstractMatrix{T}) where {T}
```
Comment (yebai) [inline src/trajectory.jl]:
```suggestion
function randcat(rng::AbstractRNG, P_unorm::AbstractMatrix{T}) where {T}
```

File: src/trajectory.jl [hunk 2]
```diff
@@ -688,18 +718,15 @@ function mh_accept_ratio(
     Hproposal::AbstractVector{<:T},
 ) where {T<:AbstractFloat}
     α = min.(one(T), exp.(Horiginal .- Hproposal))
-    accept = _rand(rng, length(Horiginal)) .< α
+    # NOTE: There is a chance that sharing the RNG over multiple
+    #       chains for accepting / rejecting might couple
+    #       the chains. We need to revisit this more rigirously 
+    #       in the future. See discussions at 
+    #       https://github.com/TuringLang/AdvancedHMC.jl/pull/166#pullrequestreview-367216534
+    accept = rand(rng, T, length(Horiginal)) .< α
```
Comment (yebai) [inline src/trajectory.jl]: This is potentially confusing since `rnd` in the argument has a type `rng::Union{AbstractRNG, AbstractVector{<:AbstractRNG}}`. We can simplify by enforcing a constraint that the user has to pass a vector for vectorised HMC, and a scalar otherwise. Any reason why we want to keep support for `rng` being an `AbstractRNG` type in vectorised HMC?

File: src/trajectory.jl [hunk 3]
```diff
@@ -227,22 +227,52 @@ samplecand(rng, τ::StaticTrajectory{EndPointTS}, h, z) = step(τ.integrator, h,
...
 function samplecand(rng, τ::StaticTrajectory{MultinomialTS}, h, z)
     zs = step(τ.integrator, h, z, τ.n_steps; res=[z for _ in 1:abs(τ.n_steps)])
     ℓws = -energy.(zs)
+    if eltype(ℓws) <: AbstractVector
+        ℓws = hcat(ℓws...)
+    end
     ℓws = ℓws .- maximum(ℓws)
-    p_unorm = exp.(ℓws)
-    return randcat(rng, zs, p_unorm / sum(p_unorm))
+    prob_unorm = exp.(ℓws)
```
Comment (yebai) [inline src/trajectory.jl]: Maybe switch to `ℓprob_unorm` here, as the log scale is numerically more stable, particularly when some candidates have a tiny weight while others have a large one.

File: src/trajectory.jl [hunk 4]
```diff
@@ -227,22 +227,52 @@ samplecand(rng, τ::StaticTrajectory{EndPointTS}, h, z) = step(τ.integrator, h,
...
+    u = rand(rng, T, size(P, 1))
+    C = cumsum(P; dims=2)
+    is = convert.(Int, vec(sum(C .< u; dims=2)))
+    return max.(is, 1)
+end
+
+randcat(rng::AbstractRNG, xs, p::AbstractVector) = xs[randcat(rng, p)]
+
+# xs is in the form of Vector{Vector} and has shape [n_steps][]
+function randcat(rng, xs, P_unorm::AbstractMatrix)
```
Comment (yebai) [inline src/trajectory.jl]: Similiar here.
```suggestion
function randcat(rng, zs::VectorOrMat{PhasePoint}, P_unorm::AbstractMatrix)
```

File: src/trajectory.jl [hunk 5]
```diff
@@ -227,22 +227,52 @@ samplecand(rng, τ::StaticTrajectory{EndPointTS}, h, z) = step(τ.integrator, h,
...
+
+function randcat(rng, P_unorm::AbstractMatrix{T}) where {T}
+    P = P_unorm ./ sum(P_unorm; dims=2)
+    u = rand(rng, T, size(P, 1))
+    C = cumsum(P; dims=2)
+    is = convert.(Int, vec(sum(C .< u; dims=2)))
+    return max.(is, 1)
+end
+
+randcat(rng::AbstractRNG, xs, p::AbstractVector) = xs[randcat(rng, p)]
```
Comment (yebai) [inline src/trajectory.jl]: Maybe change `xs` into `zs` for consistency? Also adding type annotation makes it more readable here.
```suggestion
randcat(rng::AbstractRNG, zs::Vector{PhasePoint}, p::AbstractVector) = xs[randcat(rng, p)]
```

File: src/trajectory.jl (hunk 1 above)
Comment (xukai92) [reply src/trajectory.jl]: Yes but it would be `Union{AbstractRNG, AbstractVector{<:AbstractRNG}}`

File: src/trajectory.jl (hunk 2 above)
Comment (xukai92) [reply src/trajectory.jl]: One reason is for to be compatible with `Random.seed!`, `Random.GLOBAL_RNG` is reset. In this case, `rng = GLOBAL_RNG` controls all randomness. Also I think it makes sense to control the randomness using a single RNG in general.

File: src/trajectory.jl (hunk 3 above)
Comment (xukai92) [reply src/trajectory.jl]: Good point.  Let me change all to log domain.

File: src/trajectory.jl (hunk 4 above)
Comment (xukai92) [reply src/trajectory.jl]: `zs` is always a vector of phase points. In vectorized mode, a single `z` stores all phase points for all chains for a particular step. I think this is a bit conterintuitive and I was trying to figure out a way to improve this. In particular, maybe we should all multiple steps to be sotred in a single `PhasePoint`. This would make a low of mutiple-chain/multiple-step operations easier. But I think it worth another PR with more discussions.

Comment (yebai) [reply src/trajectory.jl]: > In particular, maybe we should all multiple steps to be sotred in a single PhasePoint. 

I'm not sure, saving more than one (position, momentum) pairs in a PhasePoint would make PhasePoint and leapfrog related code complicated. Maybe there is a better approach.

Comment (xukai92) [reply src/trajectory.jl]: It will certainly make the code more complicated. But I think any solution have to do so, no matter in which manner, to achive high perfomrance.

**Comment by xukai92** [comment]
@yebai Anything else we want to include in this PR?

File: src/trajectory.jl (hunk 3 above)
Comment (yebai) [reply src/trajectory.jl]: That's true - it might still be good for consistency to always keep things in the log scale to avoid accidental mistreatments. What do you think?

File: src/trajectory.jl (hunk 4 above)
Comment (yebai) [reply src/trajectory.jl]: Perhaps we can keep this discussion for future PRs.

File: src/trajectory.jl [hunk 6]
```diff
@@ -227,8 +227,8 @@ samplecand(rng, τ::StaticTrajectory{EndPointTS}, h, z) = step(τ.integrator, h,
 
 ### Multinomial sampling from trajecory
 
-function randcat(rng::AbstractRNG, p_unorm::AbstractVector{T}) where {T}
-    p = p_unorm ./ sum(p_unorm)
+function randcat(rng::AbstractRNG, up::AbstractVector{T}) where {T}
```
Comment (yebai) [inline src/trajectory.jl]: Maybe rename `up` to `unnormlised_prob` for readability? `up` is really hard to parse...

File: src/trajectory.jl [hunk 7]
```diff
@@ -238,30 +238,30 @@ function randcat(rng::AbstractRNG, p_unorm::AbstractVector{T}) where {T}
...
     return max.(is, 1)
 end
 
-randcat(rng::AbstractRNG, xs, p::AbstractVector) = xs[randcat(rng, p)]
+randcat(rng::AbstractRNG, zs::AbstractVector{<:PhasePoint}, p::AbstractVector) = zs[randcat(rng, p)]
 
-# xs is in the form of Vector{Vector} and has shape [n_steps][]
-function randcat(rng, xs, P_unorm::AbstractMatrix)
-    x = similar(first(xs))
+# zs is in the form of Vector{PhasePoint{Matrix}} and has shape [n_steps][dim, n_chains]
```
Comment (yebai) [inline src/trajectory.jl]:
```suggestion
# zs is in the form of Vector{Matrix{PhasePoint}} and has shape [n_steps][dim, n_chains]
```

File: src/trajectory.jl (hunk 2 above)
Comment (yebai) [reply src/trajectory.jl]: > Also I think it makes sense to control the randomness using a single RNG in general.

Can we automatically generate random seeds for `n_chains` using a separate global RNG stream?  This way all randomness is controlled by a single global RNG, but still, keep RNG independent for separate chains.

File: src/trajectory.jl (hunk 6 above)
Comment (xukai92) [reply src/trajectory.jl]: Sounds good.

File: src/trajectory.jl (hunk 2 above)
Comment (xukai92) [reply src/trajectory.jl]: That's a good solution!

File: src/trajectory.jl (hunk 3 above)
Comment (xukai92) [reply src/trajectory.jl]: Yes I think it would make the code clear. Let's assume `randcat` takes an unnormalised log probability then.

Comment (xukai92) [reply src/trajectory.jl]: If so I guess we should renmae `randcat` to something else that indicates it takes probabilities in log space.

Another question is do we want to keep probability in log space even within `randcat`? If so it would requires change all accumation of probabities to take use of `logaddexp`, which I assume would be slower than simply doing addition or `cumsum`.

File: src/trajectory.jl (hunk 2 above)
Comment (xukai92) [reply src/trajectory.jl]: I took a look into how to remove all places with `Union{AbstractRNG, AbstractVector{<:AbstractRNG}}` and some of them requires some works on the metrics space part. I think this is out of the scope of this PR and worth a seperate one because it actually influences the whole code space. I created an issue (https://github.com/TuringLang/AdvancedHMC.jl/issues/186) to track this.

File: src/trajectory.jl (hunk 3 above)
Comment (yebai) [reply src/trajectory.jl]: It's ok to work with ordinary probabilities in `randcat` I think. We can revisit this later if it causes numerical instabilities.

File: src/trajectory.jl (hunk 2 above)
Comment (yebai) [reply src/trajectory.jl]: Sounds good to me.

File: src/trajectory.jl (hunk 3 above)
Comment (xukai92) [reply src/trajectory.jl]: Cool. I renamed the function which takes log probabilities as `randcatlog` to indicate the input is in log domain.

File: src/AdvancedHMC.jl [hunk 8]
```diff
@@ -26,7 +26,7 @@ function Base.randn(rng::AbstractVector{<:AbstractRNG}, T, dim::Int, n_chains::I
     return cat(randn.(rng, T, dim)...; dims=2)
 end
 
-randcatlog(rng::AbstractRNG, unnorm_ℓp::AbstractVector{T}) where {T} =
+randcatlog(rng::AbstractRNG, unnorm_ℓp::AbstractVector) =
```
Comment (yebai) [inline src/AdvancedHMC.jl]: `randcatlog` is not more readable than `randcat`... 
Maybe a docstring explaining that `randcat` assumes log probabilities is what we need here.

Comment (xukai92) [reply src/AdvancedHMC.jl]: I think this information should be reflected by the name. Basically `randcatlog` and `randcat` are two different functions. Even with a docstring it would make the code confusing when reading something like `randcat(ℓp)` but for `randcatlog(ℓp)` at least we would then realise what it is doing from the naming.

Comment (xukai92) [reply src/AdvancedHMC.jl]: Maybe there is a better name than `randcatlog`

**Comment by xukai92** [comment]
This PR is ready to merge once the naming of `randcat` is resolved. I think we should have something which reflects the probability in log domain, maybe `randlogcat` is better than `randcatlog`?

**Comment by yebai** [comment]
> This PR is ready to merge once the naming of randcat is resolved. I think we should have something which reflects the probability in log domain, maybe randlogcat is better than randcatlog?


Maybe `randcat_logp`?

**Comment by xukai92** [comment]
Sounds good. I just made the change.
